File: parsers/swift/Sources/PassportProverLib/CscaRegistry.swift
import Foundation
import Security

public struct CscaEntry {
    public let publicKey: Data
    public let subject: String
    public let serial: String
}

public enum CscaRegistry {

    public static func load(path: String) throws -> [String: [CscaEntry]] {
        let data = try Data(contentsOf: URL(fileURLWithPath: path))
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw PassportError.dataNotFound("Invalid CSCA registry JSON")
        }

        var registry: [String: [CscaEntry]] = [:]
        for (country, entriesAny) in json {
            guard let entries = entriesAny as? [[String: Any]] else { continue }
            var cscaList: [CscaEntry] = []
            for entry in entries {
                guard let pubKeyB64 = entry["public_key"] as? String,
                      let pubKeyBytes = Data(base64Encoded: pubKeyB64) else { continue }
                cscaList.append(CscaEntry(
                    publicKey: pubKeyBytes,
                    subject: entry["subject"] as? String ?? "",
                    serial: entry["serial"] as? String ?? ""
                ))
            }
            registry[country] = cscaList
        }
        return registry
    }

    public static func findMatchingKey(
        registry: [String: [CscaEntry]],
        country: String,
        dg1: Data,
        sod: SOD
    ) throws -> Data {
        guard let entries = registry[country] else {
            throw PassportError.dataNotFound("No CSCA entries for country: \(country)")
        }

        let tbsBytes = sod.certificate.tbs.bytes
        let cscaSignatureBytes = sod.certificate.signature

        var errors: [String] = []
        for entry in entries {
            do {
                if try verifyRSASignature(publicKeyDER: entry.publicKey, data: tbsBytes, signature: cscaSignatureBytes) {
                    print("  Matched CSCA key: serial=\(entry.serial)")
                    return entry.publicKey
                } else {
                    errors.append("serial=\(entry.serial): signature mismatch")
                }
            } catch {
                errors.append("serial=\(entry.serial): \(error.localizedDescription)")
            }
        }

        throw PassportError.dataNotFound(
            "No matching CSCA key found for \(country). Tried \(entries.count) keys:\n  " +
            errors.joined(separator: "\n  ")
        )
    }

    /// Verify RSA SHA256 signature using Security.framework.
    private static func verifyRSASignature(publicKeyDER: Data, data: Data, signature: Data) throws -> Bool {
        // Create SecKey from SubjectPublicKeyInfo DER
        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
        ]

        var error: Unmanaged<CFError>?
        guard let secKey = SecKeyCreateWithData(publicKeyDER as CFData, attributes as CFDictionary, &error) else {
            if let err = error?.takeRetainedValue() {
                throw PassportError.dataNotFound("Failed to create SecKey: \(err)")
            }
            throw PassportError.dataNotFound("Failed to create SecKey")
        }

        let algorithm = SecKeyAlgorithm.rsaSignatureMessagePKCS1v15SHA256
        guard SecKeyIsAlgorithmSupported(secKey, .verify, algorithm) else {
            throw PassportError.unsupportedSignatureAlgorithm("RSA SHA256 not supported by key")
        }

        return SecKeyVerifySignature(secKey, algorithm, data as CFData, signature as CFData, &error)
    }
}

File: parsers/kotlin/src/main/kotlin/verity/passport/prover/CscaRegistry.kt
package verity.passport.prover

import org.json.JSONObject
import java.io.File
import java.security.KeyFactory
import java.security.Signature
import java.security.MessageDigest
import java.security.spec.X509EncodedKeySpec
import java.util.Base64

public data class CscaEntry(
    val publicKey: ByteArray,
    val subject: String,
    val serial: String,
)

/**
 * Registry for CSCA (Country Signing Certificate Authority) public keys,
 * loaded from a JSON file and used to verify DSC certificate signatures.
 */
public object CscaRegistry {

    /**
     * Loads the CSCA registry from a JSON file at [path], returning a map
     * of ISO country codes to their list of known [CscaEntry] public keys.
     */
    public fun load(path: String): Map<String, List<CscaEntry>> {
        val json = JSONObject(File(path).readText())
        return buildMap {
            for (country in json.keys()) {
                val entries = json.getJSONArray(country)
                put(country, buildList {
                    for (i in 0 until entries.length()) {
                        val entry = entries.getJSONObject(i)
                        val pubKeyBytes = Base64.getDecoder().decode(entry.getString("public_key"))
                        add(CscaEntry(
                            publicKey = pubKeyBytes,
                            subject = entry.optString("subject", ""),
                            serial = entry.optString("serial", ""),
                        ))
                    }
                })
            }
        }
    }

    /**
     * Find the matching CSCA public key by verifying the CSCA signature
     * over the DSC TBS certificate, matching the Rust pipeline's approach.
     */
    public fun findMatchingKey(
        registry: Map<String, List<CscaEntry>>,
        country: String,
        dg1: ByteArray,
        sod: SOD,
    ): ByteArray {
        val entries = registry[country]
            ?: throw PassportError.DataNotFound("No CSCA entries for country: $country")

        // The DSC's TBS certificate bytes and signature from the SOD
        val tbsBytes = sod.certificate.tbs.bytes
        val tbsDigest = MessageDigest.getInstance("SHA-256").digest(tbsBytes)
        val cscaSignatureBytes = sod.certificate.signature

        val errors = mutableListOf<String>()
        for (entry in entries) {
            try {
                // Load the CSCA public key from SubjectPublicKeyInfo (DER)
                val keySpec = X509EncodedKeySpec(entry.publicKey)
                val keyFactory = KeyFactory.getInstance("RSA")
                val publicKey = keyFactory.generatePublic(keySpec)

                // Verify: CSCA signed the DSC TBS certificate
                val sig = Signature.getInstance("SHA256withRSA")
                sig.initVerify(publicKey)
                sig.update(tbsBytes)

                if (sig.verify(cscaSignatureBytes)) {
                    return entry.publicKey
                } else {
                    errors.add("serial=${entry.serial}: signature mismatch")
                }
            } catch (e: Exception) {
                errors.add("serial=${entry.serial}: ${e.message}")
            }
        }

        throw PassportError.DataNotFound(
            "No matching CSCA key found for $country. Tried ${entries.size} keys:\n  " +
                errors.joinToString("\n  ")
        )
    }
}
